mipengine/node/monetdb_interface/monet_db_connection.py

```diff
@@ -1,117 +1,183 @@
 from contextlib import contextmanager
-from time import sleep
+from functools import wraps
 from typing import List
 
 import pymonetdb
+from eventlet.greenthread import sleep
+from eventlet.lock import Semaphore
 
 from mipengine.node import config as node_config
 from mipengine.node import node_logger as logging
 from mipengine.singleton import Singleton
 
+INTEGRITY_ERROR_RETRY_INTERVAL = 1
 BROKEN_PIPE_MAX_ATTEMPTS = 50
-OCC_MAX_ATTEMPTS = 50
-INTEGRITY_ERROR_RETRY_INTERVAL = 0.5
+BROKEN_PIPE_ERROR_RETRY = 0.2
+CREATE_OR_REPLACE_QUERY_TIMEOUT = 1
+CREATE_REMOTE_TABLE_QUERY_TIMEOUT = 1
+INSERT_INTO_QUERY_TIMEOUT = (
+    node_config.celery.worker_concurrency * node_config.celery.run_udf_time_limit
+)
 
+create_remote_table_query_lock = Semaphore()
+create_function_query_lock = Semaphore()
+insert_query_lock = Semaphore()
 
-class MonetDB(metaclass=Singleton):
+
+class DBExecutionDTO:
+    def __init__(self, query, parameters=None, many=False):
+        self.query = query
+        self.parameters = parameters
+        self.many = many
+
+
+def query_execution_exception_handling(func):
     """
-    MonetDB is a Singleton class because we want it to be initialized at runtime.
+    On the query execution we need to handle the 'BrokenPipeError' exception.
+    In the case of the 'BrokenPipeError' exception, we create a new connection,
+    and we retry for x amount of times the execution in case the database has recovered.
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        for tries in range(BROKEN_PIPE_MAX_ATTEMPTS):
+            conn = self._get_connection()
+
+            try:
+                function = func(self, *args, **kwargs, conn=conn)
+                break
+            except BrokenPipeError as exc:
+                conn = self._create_connection()
+                sleep(tries * BROKEN_PIPE_ERROR_RETRY)
+                continue
+            except Exception as exc:
+                conn.rollback()
+                raise exc
+            finally:
+                self._release_connection(conn)
+        else:
+            raise exc
+
+        return function
 
-    If the connection is a public module variable, it will be initialized at import time
-    from Celery and all the Celery workers will use the same connection instance.
+    return wrapper
 
-    We want one MonetDB connection instance per Celery worker/process.
 
+class MonetDBPool(metaclass=Singleton):
+    """
+    MonetDBPool is a Singleton class because we want it to be initialized at runtime.
+
+    We use sudo-multithreading(eventlet greenlets),
+    we provide a connection pool to support concurrent query execution.
```

Comment:
What is the initialization you re referring to? Also the goal when implementing a class as Singleton is to allow only one instance of the class to exist at any time, what does " because we want it to be initialized at runtime" mean?
I suppose by "sudo-multithreading" you mean pseudo-multithreading?
Naming `MonetDBPool` indicates a pool of multiple MonetDBs and not a pool of multiple connections

Reply:
Yes I was meaning pseudo-multithreading.
You are right maybe MonetDBConnectionPool would be more appropriate.